File: src/chipschallenge/blockreactions/TrapTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;

public class TrapTo extends NoSlipReaction {

    private TrapTo() {
    }
    private static TrapTo mInstance = null;

    public static synchronized TrapTo getInstance() {
        if (mInstance == null) {
            mInstance = new TrapTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        if(moving.isBlock() && moving.isForced()) {
            game().addForcedMove(moving, moving.getFacing());
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/TrapTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class TrapTo : NoSlipReaction
    {
        private static TrapTo instance;

        private TrapTo()
        {
        }

        public static TrapTo Instance
        {
            get
            {
                lock (typeof(TrapTo))
                {
                    return instance ?? (instance = new TrapTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.IsBlock() && moving.Forced)
            {
                Game().AddToSlipList(moving, moving.Facing);
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}